test/features/build_server/cache/project_cache_test.dart

```diff
@@ -9,7 +16,72 @@ import '../../../helpers/create_dummy_project_cache.dart';
 const _cacheId = 'project_cache_test';
 const _dummyProjectCachePath = './flyde-test-lib-$_cacheId/cache/testing';
 
+/// Reads the `./example` directory and returns a list of all files.
+Future<List<File>> _loadExampleFiles() async {
+  return [
+    await for (final entity in Directory('./example').list(recursive: true)) entity,
+  ]
+      .whereType<File>()
+      .where((file) =>
+          FileExtension.sources.contains(p.extension(file.path)) ||
+          FileExtension.headers.contains(p.extension(file.path)))
+      .toList();
+}
+
+/// Returns a list of all files in the `./example` directory converted
+/// to `SourceFile`s.
+///
+/// If [changing] is not null, all files with the this name will be
+/// changed using the [withContent] function on it's data.
+Future<List<SourceFile>> _getSourceFiles({
+  String? changing,
+  String Function(String)? withContent,
+}) async {
+  final files = await _loadExampleFiles();
+  final srcFiles = <SourceFile>[];
+
+  for (final file in files) {
+    final srcFile = SourceFile.fromFile(0, file, entryDirectory: Directory('./example'));
+
+    if (srcFile.name == changing && withContent != null) {
+      final newCnt = withContent(utf8.decode(await srcFile.data));
+      srcFiles.add(SourceFile(
+        srcFile.entry,
+        srcFile.path,
+        srcFile.name,
+        srcFile.extension,
+        data: Uint8List.fromList(utf8.encode(newCnt)),
+      ));
+    } else {
+      srcFiles.add(srcFile);
+    }
+  }
+
+  return srcFiles;
+}
+
+/// Returns a list of all required file ids of [cache] with the
+/// given [config] after synchronizing with [srcFiles].
+Future<List<String>> _getRequiredFiles(
+  ProjectCache cache,
+  CompilerConfig config,
+  List<SourceFile> srcFiles,
+) async {
+  final fileMap = {
+    for (final file in srcFiles) file.id: await file.hash,
+  };
+
+  return await cache.sync(fileMap, config);
+}
+
 void main() {
+  final config = CompilerConfig(
+      compiler: InstalledCompiler.gpp,
+      threads: 4,
```

Comment:
Maybe a bit high for test runners?

Reply:
I'd like to keep it at 4 to be able to test multi threaded compilation properly and because the example project is not meant to grow very big. It should not make any difference how many threads we are using.

Reply:
Sounds reasonable